numaprom/udf/postprocess.py

```diff
@@ -16,115 +19,125 @@
 AUTH = os.getenv("REDIS_AUTH")
 
 
-def save_to_redis(payload: Payload, recreate: bool):
+def save_to_redis(payload: StreamPayload, final_score: float, recreate: bool):
     r = get_redis_client(HOST, PORT, password=AUTH, recreate=recreate)
 
-    metric_config = get_metric_config(payload.metric_name)
+    metric_name = payload.composite_keys["name"]
+    metric_config = get_metric_config(metric_name)
     model_config = metric_config["model_config"]
 
     metrics_list = model_config["metrics"]
-    key = ":".join(metric_config["keys"])
-    key = f"{key}:{payload.endTS}"
+    r_key = f"{':'.join(metric_config['keys'])}:{payload.end_ts}"
 
-    if np.isnan(payload.anomaly):
-        r.hset(key, mapping={payload.metric_name: -1})
+    if np.isnan(final_score):
+        r.hset(r_key, mapping={metric_name: -1})
     else:
-        r.hset(key, mapping={payload.metric_name: payload.anomaly})
+        r.hset(r_key, mapping={metric_name: final_score})
 
     for m in metrics_list:
-        if not r.hexists(name=key, key=m):
+        if not r.hexists(name=r_key, key=m):
             return -1, []
 
     max_anomaly = -1
     anomalies = []
     for m in metrics_list:
-        val = float(r.hget(name=key, key=m))
+        val = float(r.hget(name=r_key, key=m))
         anomalies.append(val)
         if max_anomaly < val:
             max_anomaly = val
-    r.delete(key)
+    r.delete(r_key)
 
     return max_anomaly, anomalies
 
 
-def get_labels(payload: Payload):
-    metric_config = get_metric_config(payload.metric_name)
-    keys = metric_config["keys"]
-    labels = {"model_version": str(payload.model_version)}
-    for key in keys:
-        if key != "name":
-            labels[key] = payload.src_labels[key]
-
-    return labels
+def __construct_publisher_payload(
+    stream_payload: StreamPayload, final_score: float
+) -> PrometheusPayload:
+    metric_name = stream_payload.composite_keys["name"]
+    namespace = stream_payload.composite_keys["namespace"]
 
+    labels = {
+        "model_version": str(stream_payload.get_metadata("version")),
+        **stream_payload.get_metadata("src_labels"),
```

Comment:
do not add all the `src_labels`, add only what is required as it could lead to high cardinality data. And, especially in unified anomaly, it could take src_labels from a different metric each time, and will be regarded as a new time series. 

> CAUTION: Remember that every unique combination of key-value label pairs represents a new time series, which can dramatically increase the amount of data stored. Do not use labels to store dimensions with high cardinality (many different label values), such as user IDs, email addresses, or other unbounded sets of values.

Reply:
Add hash_id as a label for sure

Reply:
Thanks for this. Can we define, what all labels do we need? It will help us to create a structure, and not be surprised :)

Reply:
Ideally, it should just be the ones from composite keys excluding the metric name. So, just namespace for argo cd anomaly, namespace and hash_id for rollouts anomaly. 

If the users want some extra labels to be added, we could add it to metric config. (in the future)